triplea-game/triplea
# Add canal validation to route finding

## Overview
With the release of RSOC v3, the lack of canal validation during route finding is very apparent. This has been true for all maps but most have only a handful of canals where RSOC has hundreds. Route finding should do some basic canal checks to ensure the canal is controlled and at least 1 of the selected units can move through.

## Functional Changes
- Route finding now does some basic canal validation to ensure at least 1 of the selected units can pass through the canal in order to consider using it on the route
- Canal attachment validation moved to map parsing rather than route finding...
- Improved performance of checking if a canal is on a route by avoiding looping through every territory in the map

## Manual Testing Performed
- Tested mostly on RSOC but checked a few standard maps like revised and global to ensure no issues
- Performance was initially awful but after a couple iterations should now seem to be the same as before to regular users
- Tested editing a map XML to have only 1 of the pair of canal attachments specified to ensure the error is caught during map parsing

## Before & After Screen Shots

### Before
![image](https://user-images.githubusercontent.com/1427689/45337777-581adf00-b54f-11e8-9740-644ef2badfef.png)
![image](https://user-images.githubusercontent.com/1427689/45338010-ab416180-b550-11e8-85f5-821b35251913.png)


### After
![image](https://user-images.githubusercontent.com/1427689/45337803-75e84400-b54f-11e8-85e9-38a162b89c3b.png)
![image](https://user-images.githubusercontent.com/1427689/45337971-7208f180-b550-11e8-9999-b7b369da0a97.png)
